zkvm/src/prover.rs

```diff
@@ -39,13 +54,83 @@ impl<'a, 'b> Delegate<r1cs::Prover<'a, 'b>> for Prover<'a, 'b> {
         }
     }
 
+    fn next_instruction(
+        &mut self,
+        run: &mut Self::RunType,
+    ) -> Result<Option<Instruction>, VMError> {
+        let instruction = run.program.pop_front();
+        if run.root {
+            match &instruction {
+                Some(i) => i.encode(&mut self.bytecode),
+                None => (),
+            };
+        }
+        Ok(instruction)
+    }
+
     fn cs(&mut self) -> &mut r1cs::Prover<'a, 'b> {
         &mut self.cs
     }
 }
 
-impl RunTrait for RunProver {
-    fn next_instruction(&mut self) -> Result<Option<Instruction>, VMError> {
-        unimplemented!()
+impl<'a, 'b> Prover<'a, 'b> {
+    pub fn build_tx<'g>(
+        program: Vec<Instruction>,
+        version: u64,
+        mintime: u64,
+        maxtime: u64,
+        bp_gens: &'g BulletproofGens,
+    ) -> Result<Tx, VMError> {
+        let mut r1cs_transcript = Transcript::new(b"ZkVM.r1cs");
+        let pc_gens = PedersenGens::default();
+        let cs = r1cs::Prover::new(bp_gens, &pc_gens, &mut r1cs_transcript);
+
+        let mut prover = Prover {
+            signtx_keys: Vec::new(),
+            cs: cs,
+            bytecode: Vec::new(),
+        };
+
+        let vm = VM::new(
+            version,
+            mintime,
+            maxtime,
+            ProverRun::from_txprogram(program),
+            &mut prover,
+        );
+
+        let (txid, _) = vm.run()?;
```

Comment:
Maybe we should return txlog together with `Tx`, for convenience? Then the prover can avoiding passing the Tx through the Verifier to get it.

`build_tx() -> Result<(Tx,Vec<txlog::Entry>), VMError>`